bpf.go

```diff
@@ -317,15 +323,134 @@ func constructFilter(subsys Table, filters []ConnAttr) ([]bpf.RawInstruction, er
 	return raw, nil
 }
 
-func (nfct *Nfct) attachFilter(subsys Table, filters []ConnAttr) error {
+func filterMarkAttribute(filters []ConnAttr) []bpf.RawInstruction {
+	var raw []bpf.RawInstruction
+
+	// sizeof(nlmsghdr) + sizeof(nfgenmsg) = 20
+	tmp := bpf.RawInstruction{Op: unix.BPF_LD | unix.BPF_IMM, K: 0x14}
+	raw = append(raw, tmp)
+
+	// find final attribute
+	tmp = bpf.RawInstruction{Op: unix.BPF_LDX | unix.BPF_IMM, K: uint32(filterCheck[filters[0].Type].ct)}
+	raw = append(raw, tmp)
+	tmp = bpf.RawInstruction{Op: unix.BPF_LD | unix.BPF_B | unix.BPF_ABS, K: 0xfffff00c}
+	raw = append(raw, tmp)
+
+	tmp = bpf.RawInstruction{Op: unix.BPF_JMP | unix.BPF_JEQ | unix.BPF_K, K: 0, Jt: 2}
+	raw = append(raw, tmp)
+
+	tmp = bpf.RawInstruction{Op: unix.BPF_MISC | unix.BPF_TAX}
+	raw = append(raw, tmp)
+
+	tmp = bpf.RawInstruction{Op: unix.BPF_LD | unix.BPF_W | unix.BPF_IND, K: uint32(len(filters[0].Data))}
+	raw = append(raw, tmp)
+
+	tmp = bpf.RawInstruction{Op: unix.BPF_MISC | unix.BPF_TAX}
+	raw = append(raw, tmp)
+
+	for _, filter := range filters {
+		var dataLen = len(filter.Data)
+		for i := 0; i < (int(dataLen) / 4); i++ {
+			mask := encodeValue(filter.Mask[i*4 : (i+1)*4])
+			tmp = bpf.RawInstruction{Op: unix.BPF_ALU | unix.BPF_AND | unix.BPF_K, K: mask}
+			raw = append(raw, tmp)
+			val := encodeValue(filter.Data[i*4 : (i+1)*4])
+			val &= mask
+			tmp = bpf.RawInstruction{Op: unix.BPF_JMP | unix.BPF_JEQ | unix.BPF_K, K: val, Jt: failedJump}
+			raw = append(raw, tmp)
+			tmp = bpf.RawInstruction{Op: unix.BPF_MISC | unix.BPF_TXA}
+			raw = append(raw, tmp)
+		}
+	}
+
+	var j uint8 = 1
+
+	// Failed jumps are set to 255. Now we correct them to the actual failed jump instruction
+	for i := len(raw) - 1; i > 0; i-- {
+		if (raw[i].Jt == failedJump) && (raw[i].Op == unix.BPF_JMP|unix.BPF_JEQ|unix.BPF_K) {
+			raw[i].Jt = j
+		}
+		j++
+	}
 
+	// negate filter
+	if filters[0].Negate {
+		raw = append(raw, bpf.RawInstruction{Op: unix.BPF_JMP | unix.BPF_JA, K: 1})
+	}
+
+	// reject
+	raw = append(raw, bpf.RawInstruction{Op: unix.BPF_RET | unix.BPF_K, K: bpfVerdictReject})
+
+	return raw
+}
+
+func (nfct *Nfct) attachFilter(subsys Table, filters []ConnAttr) error {
 	bpfFilters, err := constructFilter(subsys, filters)
 	if err != nil {
 		return err
 	}
+	if nfct.debug {
+		fmtInstructions := fmtRawInstructions(bpfFilters)
+		fmt.Println("---BPF filter start---")
+		fmt.Print(fmtInstructions)
+		fmt.Println("---BPF filter end---")
```

Comment:
It should be up to the user, where to write down information. Therefore, the `logger` element exists in `nfct`.
```suggestion
		nfct.logger.Println("---BPF filter start---")
		nfct.logger.Print(fmtInstructions)
		nfct.logger.Println("---BPF filter end---")
```

Reply:
Sure, good catch